Example Error Objects

bf4 opened the pull request:
Per discussion with @dgeb in https://github.com/rails-api/active_model_serializers/pull/1004

> Although it is still compliant to represent multiple validation errors in the same error object,
> it's not ideal. I would recommend creating a unique error object per validation error,
> so that fields such as code and title can be used consistently and expressively across errors.
> For instance, one code might indicate that a field is too short, and another might indicate
> that invalid characters were used. If you try to combine these errors into a single code,
> then it will probably end up being something more generic with less meaning, such as "invalid value".

References:
- http://jsonapi.org/format/#error-object
- Discussion in PRs
  - https://github.com/json-api/json-api/pull/237 [specifically](https://github.com/json-api/json-api/commit/6901143531c14f41e95818455fe36228cebf6625) and https://github.com/json-api/json-api/pull/638
  - https://github.com/json-api/json-api/pull/234
  - https://github.com/json-api/json-api/issues/7
  - https://github.com/json-api/json-api/pull/208
- Usage in Ember
  - http://emberjs.com/blog/2015/06/18/ember-data-1-13-released.html#toc_using-json-api-error-object-format
  - ember-data
    - [lib/adapters/errors.js](https://github.com/emberjs/data/blob/d842890357840700f0ee17f152415be0cba16030/packages/ember-data/lib/adapters/errors.js#L114-L132)
    - [lib/adapters/rest-adapter.js](https://github.com/emberjs/data/blob/3930e457f36a0bbdde0155c1b98ebe976e607358/packages/ember-data/lib/adapters/rest-adapter.js#L913-L925)
    - [lib/serializers/json-serializer.js](https://github.com/emberjs/data/blob/944f7e4b537cdc7cd2a80718d0447453c01de2db/packages/ember-data/lib/serializers/json-serializer.js#L1225-L1276)
    - [adapter-errors-test.js](https://github.com/emberjs/data/blob/21ebe9b61b33acd910c0d485e17926c89c51d9f6/packages/ember-data/tests/unit/adapter-errors-test.js#L32-L48)
    - [json-serializer-test.js](https://github.com/emberjs/data/blob/90a0011c8e5e1c41d1667c96a38c4a5e4a9c1e64/packages/ember-data/tests/integration/serializers/json-serializer-test.js#L566-L622)

dgeb commented on examples/index.md:
```diff
@@ -126,3 +126,119 @@ Content-Type: application/vnd.api+json
+{
+  "errors": [
+    {
+      "source": { "pointer": "data/attributes/first_name" },
```
Although underscores are compliant, I'd prefer to use dasherized names, such as `first-name`, in examples as per our recommendations.

steveklabnik commented:
Thanks for this! Good examples are so crucial to getting a spec adopted.

dgeb commented on examples/index.md:
```diff
@@ -126,3 +126,119 @@ Content-Type: application/vnd.api+json
+      "detail": "is too short"
+    },
+    {
+      "status": "422",
```
Having `status` in isolation seems a bit odd. In this example, I'd expect it to be part of the first error object.

I think an example that shows multiple different errors with different `status` values would be great. For example, one error object could be a `404` and another `422`, and the status for the whole response could be `400`.

bf4 replied:
Great, thanks!  I made it this way on purpose since I didn't understand when error objects might have different status attributes.

dgeb commented on examples/index.md:
```diff
@@ -126,3 +126,119 @@ Content-Type: application/vnd.api+json
+      "code":   "123",
+      "title":  "too short",
+      "source": { "pointer": "data/attributes/first_name" },
+      "detail": "is too short"
```
`detail` could be longer than `title` and less of a fragment (the spec also says "specific to this occurrence of the problem."). Perhaps `"First name must contain at least three characters."` for this example.

bf4 replied:
Thanks.  I'm glad you caught that :)

dgeb commented:
> Thanks for this! Good examples are so crucial to getting a spec adopted.

Agreed! Great start here, @bf4 :+1:

bf4 commented:
I've made some updates, there might be some code blindness going on here.. I'm not sure if I did 'parameter' right. It appears to have evolved from 'path'.

I'm also thinking about adding examples of an error object with
- a links object
- a meta object
- a source pointer to a 'primary' data object.

And a response with, in addition to an `errors` top-level member, the other valid top-level members
- meta: a meta object that contains non-standard meta-information.
- jsonapi: an object describing the server's implementation
- links: a links object related to the primary data. (?)

bf4 commented:
@dgeb @steveklabnik Any more feedback on the changes? (I've made them in incremental commits)

dgeb commented on examples/index.md:
```diff
@@ -126,3 +126,206 @@ Content-Type: application/vnd.api+json
+{
+  "errors": [
+    {
+      "id": "ds.errors.invalid-error-expects-json-api-format",
```
This looks more like a `code` than an `id`:

> id: a unique identifier for this particular occurrence of the problem.
> code: an application-specific error code, expressed as a string value.

bf4 replied:
Oh, looks like I somehow mixed up code and id here and in the error codes section as well

FWIW, I got it from https://github.com/emberjs/data/blob/d842890357840700f0ee17f152415be0cba16030/packages/ember-data/lib/adapters/errors.js#L82

``` es6
    Ember.deprecate('`InvalidError` expects json-api formatted errors.', false, { id: 'ds.errors.invalid-error-expects-json-api-format', until: '2.0.0' });
```

which I misread as an errors object

bf4 replied:
Updated

dgeb commented on examples/index.md:
````diff
@@ -162,3 +162,209 @@ Content-Type: application/vnd.api+json
+```
+
+```http
+HTTP/1.1 404 Not Found
````
Should be `400`, since errors include `404` and `403` statuses.

dgeb commented on examples/index.md:
```diff
@@ -162,3 +162,209 @@ Content-Type: application/vnd.api+json
+{
+  "errors": [
+    {
+      "id":   "123",
```
These really are codes as well, if they match up with the table above. The point is that `id` represents a specific instance of an error, that may be trackable in a database for instance. If the same `id` could be reused across requests, it's really a `code`.

bf4 replied:
Updated. I'll rebase if it looks good. 

So, 'id' could be a uuid generated at error time and logged as part of that request/response. Gotcha.

bf4 commented on examples/index.md:
````diff
@@ -162,3 +162,209 @@ Content-Type: application/vnd.api+json
+
+```http
+HTTP/1.1 422 Unprocessable Entity
+Content-Type: application/vnd.api+json
````
400 Bad Request?

dgeb replied:
> 400 Bad Request?

:+1:

bf4 commented on examples/index.md:
```diff
@@ -162,3 +162,226 @@ Content-Type: application/vnd.api+json
+      "source": { "pointer": "data/relationships/author" },
+      "detail": "Post author cannot be blank."
+    }
+  ]
```
refs in the wild: 
- https://github.com/cerebris/jsonapi-resources/blob/28ac3971f8b9e73073575d41439ffe263d7e56f5/test/controllers/controller_test.rb#L452-L462

bf4 commented on examples/index.md:
```diff
@@ -162,3 +162,226 @@ Content-Type: application/vnd.api+json
+      "title":  "Invalid Attribute",
+      "source": { "pointer": "data/attributes/first-name" },
+      "detail": "First name must contain at least three characters."
+    }
```
refs in the wild: 
- https://github.com/cerebris/jsonapi-resources/blob/28ac3971f8b9e73073575d41439ffe263d7e56f5/test/controllers/controller_test.rb#L2412-L2429
- https://github.com/chingor13/json_api_client/blob/cf79aba96cbc9ff66a793d07f43b025fd8f70b38/lib/json_api_client/error_collector.rb#L34-L48
- https://github.com/manyminds/api2go/blob/3023e50d563e6ec642b46a4f044812cb232661a1/error.go
- https://github.com/emberjs/data/blob/1f866da3b4a4edcf70fdf346cd6e6be6cfaecfab/packages/ember-data/lib/adapters/errors.js#L3-L154

bf4 commented on examples/index.md:
```diff
@@ -162,3 +162,249 @@ Content-Type: application/vnd.api+json
+    {
+      "title":  "Missing Query Parameter",
+      "source": { "parameter": "data" },
+      "detail": "The required parameter, data, is missing."
```
I've found examples of these in the wild, but they don't appear to adhere to the spec.  I chose this example because I would have thought a POST with the data member mispelled, and thus missing, wouldn't count as a missing query parameter http://jsonapi.org/format/#query-parameters

This could also have been `"source": { "pointer": "/data" }`, right?

ethanresnick replied:
As I understand it, "query parameter" in the spec is referring strictly to query parameters in the request uri (like `sort`, `fields`, etc.), not to parameters in a broader sense (e.g. payload data).So a GET request that included a `fields[TYPE]` parameter whose value specified a field that doesn't exist on resources of that type would be an example of when to use the `parameter` member in the error response.

bf4 replied:
@ethanresnick Thanks for confirming my suspicions.  I glanced through some docs on the web and didn't immediately find anything that distinguished url params from body params when discussing query params.

And I suppose it would be up the implementation to decide whether to ignore an invalid or missing parameter or to return an error.

I think I can keep this example as a POST with body `{ "datum": { } }` to return an error object with `"source": { "pointer": "/data" }`  and add a new one for parameter, but I can't think of an  example where it wouldn't make sense for the client to ignore an invalid or missing parameter.

`GET http://www.example.com/api/users/1/secrets.json`  could return a `missing auth_key parameter`?

ethanresnick replied:
> GET http://www.example.com/api/users/1/secrets.json could return a missing auth_key parameter?

That would be fine. I think returning an error (rather than ignoring) an invalid `fields` or `include` parameter would be fine too, but perhaps your example is safer.

bf4 replied:
I was thinking an implementor might decide to ignore invalid params that aren't required, so it wouldn't be as strong an example.  Does that jive with what you're thinking?

ethanresnick replied:
Yup

bf4 commented:
@dgeb updated and rebased and added a few more source examples.   What would a meta example be in an error object (not top-level member meta)?

bf4 commented:
New and improved :sparkles:

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,292 @@ Content-Type: application/vnd.api+json
+    {
+      "title": "Adapter Error",
+      "code": "ds.errors.invalid-error-expects-json-api-format",
+      "detail": "The adapter rejected the commit because it was invalid"
```
Can you make the detail message here a bit more distinct than the `title` than it is now?

Also, can you add `status` to one of the earlier examples, in addition to having it on the one below? I don't want to imply through the examples that `status` is only applicable in cases where there are multiple errors returned at once. It can be useful too when the code processing the JSON isn't directly aware of the HTTP response (and could be useful also if JSON API is eventually used over non-HTTP protocols).

bf4 replied:
Sure.  FWIW, I got that example title/code/detail off of the ember-data project :).

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,292 @@ Content-Type: application/vnd.api+json
+      "detail": "Access to secret powers not authorized on Sundays."
+    },
+    {
+      "status": "404",
```
The `status` for this error object should be `400`, I think. Because 404 only applies if the resource isn't found, but the `vorpal-blade` attribute isn't it's own resource; it's just (an invalid) part of a bigger resource that _was_ found.

bf4 replied:
Hmm, I wanted a 404 example.  I'll update this. I intended this to reference a link, not an attribute.

bf4 replied:
Updated to 

``` json
{
      "status": "404",
      "links": { "about": {
        "href": "http://example.com/weapons/vorpal-blade",
        },
      },
      "title": "Weapon not found."
}
```

I'm a little unclear on how the error links object should work.

bf4 replied:
I think this could also have been, instead of links,  a `"source": { "pointer": "data/relationships/weapons"}`

ethanresnick commented on examples/index.md:
````diff
@@ -162,3 +162,292 @@ Content-Type: application/vnd.api+json
+Multiple errors on `first-name` attribute:
+
+```http
+HTTP/1.1 422 Unprocessable Entity
````
422 works here, but 400 is [also valid](http://stackoverflow.com/a/20215807/1261879), and some people still don't like using status codes from WebDAV for non-webDAV apps.

Therefore, maybe you could add a line at the bottom of this example saying that a 400 response would also be ok, and linking to that SO answer? I don't want JSON API to appear to be taking a position on 400 vs. 422 by only using one in our examples.

bf4 replied:
Thanks. I admit I was unsure of which to choose in some of these examples.

ethanresnick commented on examples/index.md:
````diff
@@ -162,3 +162,292 @@ Content-Type: application/vnd.api+json
+```
+
+```http
+HTTP/1.1 422 Bad Request
````
Typo; this should be 422 Unprocessable Entity

bf4 replied:
hangs head in shame :frowning:

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,292 @@ Content-Type: application/vnd.api+json
+    {
+      "title":  "Missing Data Member",
+      "source": { "source": "/data" },
+      "detail": "The required parameter, data, is missing."
```
Can you get rid of `detail` here? It basically restates the title, so it isn't needed. More importantly though, I don't want to call `data` a "parameter" since that further blurs the distinction between `"source"` and `"parameter"`; thus far, the only thing we've called "parameters" are the URI query params.

ethanresnick commented on examples/index.md:
````diff
@@ -162,3 +162,292 @@ Content-Type: application/vnd.api+json
+Basic request:
+
+```http
+GET /posts/1,2,3,4
````
`/posts/1,2,3,4` is not a URI format that we recommend any more because (esp. with UUIDs) the URI length can somewhat quickly exceed the maximum length that many browsers/servers support. Personally, I think we should still be recommending this format for many cases (because it works, and offers a performance boost, 80% of the time). But, unless we add it back as a recommendation, I don't think we should have an example that uses it. 

Also, since `parameter` just refers to URI query parameters in the RFC 3986 sense, treating `/1,2,3,4` as a source of `"parameter"` errors is invalid.

There's probably a way to rework this example with a new URI to make it compliant, but maybe we should just get rid of it? After all, you already have a very good description of `"parameter"` in the next example.

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,292 @@ Content-Type: application/vnd.api+json
+(invalid paramter key, perhaps a typo), "?redirect_to=http%3A%2F%2Fwww.owasp.com" (invalid
+parameter, in this case, a phishing attack), etc.
+
+Response with one error object, with parameter error:
```
Dangling line.

ethanresnick commented:
Hey @bf4,

Thanks again for this! I just had a chance to review the whole PR from my computer, and I've added inline comments for some final things that I think need tweaking. Once those are fixed, I think this'll be good to go!

bf4 commented:
@ethanresnick Updated, though I admit I'm getting a little code blindness.  Let me know if I should move the extension examples and the 'links: about' example into another pr if the rest could be merged sooner.

ethanresnick commented:
@bf4 Awesome, this is getting close now. Can you squash these commits down to one? Then, do you mind if I send a PR to your branch with a couple small, outstanding fixes—I think that's easier than asking you to do another round of revisions—and then we'll merge this?

bf4 commented:
Great, will do

swalkinshaw commented on examples/index.md:
Is the 2nd `source` a mistake? Spec only mentions `pointer` and `parameter`. There's another occurrence of this in the very last example if so.

bf4 commented on examples/index.md:
Good catch! It should be pointer :)

bf4 commented:
@ethanresnick Rebased and updated with @swalkinshaw's catch where I had `source: { source: "" }` instead of `{source: {pointer: ""}`. I also changed all the the words I had in quotes to backticks to be more consistent and because they show up better.  Thanks to all!

Other thoughts:
- I don't really like how I've separated each example, but didn't prefer h4 nor hr nor bold, so left it for now.
- I think the two extensions examples should go on their respective pages, and be referenced here, rather than inline.
- It could probably do examples with an "id", and for DELETE
- It doesn't really address included relations or nested relations

bf4 committed:
Example Error Objects

ethanresnick committed:
Clean up error object examples

bf4 committed:
Merge pull request #1 from TechAtNYU/errors_example

Clean up error object examples

bf4 commented on examples/index.md:
```diff
@@ -162,3 +162,241 @@ Content-Type: application/vnd.api+json
+
+The only uniqueness constraint on error objects is the `id` field. Thus,
+multiple errors on the same attribute can each be given their own error
+object. The example below shows multiple errors on the `first-name` attribute:
```
I believe we're formatting attribute names as `"first-name"`, so, note to self :)

bf4 commented on examples/index.md:
```diff
@@ -162,3 +162,241 @@ Content-Type: application/vnd.api+json
+
+> Note: in the responses above with a 422 status code, `400 Bad Request` would
+also be acceptable. ([http://stackoverflow.com/a/20215807/1261879](More details.))
+JSON API doesn't take a position on 400 vs. 422.
```
extraneous `>`

ethanresnick replied:
This is there to mark the section as an indented note. Cf the note above that says "The above example URI shows unencoded [ and ]..."

bf4 replied:
yeah.. noticed that.. 'noted'

bf4 commented on examples/index.md:
```diff
@@ -162,3 +162,241 @@ Content-Type: application/vnd.api+json
+> |  226 | Passwords do not match                                    |
+> |  227 | Password cannot be one of last five passwords             |
+
+Multiple errors on `password` attribute, with error `code`:
```
See above, should be `"password"` to be consistent with attribute quoting (note to self)

ethanresnick replied:
Yes, quoting the attribute values would be more consistent with the base spec. There, I believe we actually quote _all_ member values. So, probably, we should be quoting `code`, `detail`, `title`, etc. Honestly, though, it's not a big deal (to me anyway) one way or the other.

bf4 commented on examples/index.md:
```diff
@@ -162,3 +162,241 @@ Content-Type: application/vnd.api+json
+an invalid value for a JSON API–defined query parameter. However, for API-specific
+query parameters (i.e. those not defined by JSON API), a server may choose to
+ignore an invalid parameter and have the request succeed, rather than respond with
+an error. API-specific query paramters must contain one non a-z character.
```
typo `parameter`, also, on first read, it's unclear to me what it means for API-specific query parameters to contain a non a-z character and I haven't yet found it in the `schema` or the format page.

? http://jsonapi.org/format/#document-member-names

> To enable an easy mapping of member names to URLs, it is RECOMMENDED that member names use only non-reserved, URL safe characters specified in RFC 3986.

ethanresnick replied:
Thanks for catching the typo! Can you fix that, so I don't need to make a new PR?

Re api specific query parameters, here's the reference:

> "Implementation specific query parameters MUST adhere to the same constraints as member names with the additional requirement that they MUST contain at least one non a-z character (U+0061 to U+007A)."

Then, later:

> "This is to preserve the ability of JSON API to make additive additions to standard query parameters without conflicting with existing implementations."

bf4 replied:
Ok, thanks, found it: http://jsonapi.org/format/#query-parameters  Don't know why I didn't see it.

What would be an example of this?  `auth_key=` is a valid api-specific query param but  `authkey=` is not?

ethanresnick replied:
> What would be an example of this? auth_key= is a valid api-specific query param but authkey= is not?

Correct!

bf4 commented:
@ethanresnick I amended the merge commit with those small changes.

Good for a final review by dgeb steveklabnik etc?

I'm bringing down some notes about future work from above:

> I think the two extensions examples should go on their respective pages, and be referenced here, rather than inline.

That'll be a new PR I'll make based on the excised material

> It could probably do examples with an "id", and for DELETE

I'll make another PR if there's interest

> It doesn't really address included relations or nested relations

I'm not sure what that would look like offhand. 
- I'd also like to add a proper `links` attribute as discussed in https://github.com/bf4/json-api/pull/1#discussion_r37645336
- Maybe add a note about handling the top level member `json-api` (which should be `jsonapi`) per https://github.com/bf4/json-api/pull/1#discussion_r37646127

bf4 commented on examples/index.md:
```diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+query parameters (i.e. those not defined by JSON API), a server may choose to
+ignore an invalid parameter and have the request succeed, rather than respond with
+an error. [API-specific query parameters must contain one non a-z
+character.](http://jsonapi.org/format/#query-parameters)
```
I added this link to the merge commit

ethanresnick replied:
:+1:

bf4 commented:
btw, I'm on the [#jsonapi channel](irc://irc.freenode.net/#jsonapi)

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+  "errors": [
+    {
+      "status": "403",
+      "source": { "pointer": "data/attributes/secret-powers" },
```
Pointer should start with a `/`

bf4 replied:
oy

bf4 replied:
It would be really great were there an easy way to validate example code against the schema.  That easiest right now would be a task that parses fenced code blocks, identifies the document, and then can validate that..

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+    },
+    {
+      "status": "422",
+      "source": { "pointer": "data/attributes/volume" },
```
Ditto

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+    },
+    {
+      "status": "500",
+      "source": { "pointer": "data/attributes/reputation" },
```
Ditto

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+{
+  "errors": [
+    {
+      "source": { "pointer": "data/attributes/first-name" },
```
Ditto

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+      "detail": "First name must contain at least three characters."
+    },
+    {
+      "source": { "pointer": "data/attributes/first-name" },
```
Ditto

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+  "errors": [
+    {
+      "code":   "123",
+      "source": { "pointer": "data/attributes/first-name" },
```
Ditto

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+    },
+    {
+      "code":   "225",
+      "source": { "pointer": "data/attributes/password" },
```
Ditto

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+    },
+    {
+      "code":   "226",
+      "source": { "pointer": "data/attributes/password" },
```
Ditto

ethanresnick commented on examples/index.md:
```diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+{
+  "errors": [
+    {
+      "source": { "pointer": "/" },
```
My bad. This pointer should actually be the empty string `""`. Using `"/"` refers to a property `""` at the top level, i.e.

``` js
{
  "": "some value"    // The json pointer "/" would reference the string "some value"
}
```

ethanresnick commented:
@bf4 Thanks for all your hard work! This looks great to me now.

Please fix the small errors with the JSON Pointers, and then I think this is good to merge, unless @dgeb or @tkellen want to review it first.

bf4 commented:
Merge commit amended and force-pushed.

ethanresnick commented on examples/index.md:
````diff
@@ -162,3 +162,242 @@ Content-Type: application/vnd.api+json
+}
+```
+
+It uses `source` to point to the top-level of the document (`"/"`).
````
This line needs to say `""` instead of the old `"/"`. 

Also, I think my comment about why `"/"` is inappropriate got accidentally merged into the response. It doesn't make sense there, though, because my example document with `{"": "some value"}` is missing, and because a server isn't allowed to respond with comments in its JSON. If you want to explain why `""` is used rather than `"/"`, please do it in this paragraph instead.

bf4 replied:
you a copy editor or I'm just tired of re-reading this ? :)

ethanresnick replied:
Maybe both :)

bf4 replied:
Added an incremental commit to address this to aid in future reviewing. I'll rebase into the merge commit once we're happy.

ethanresnick commented:
:+1:

bf4 commented:
fixing merge conflict... hold on

bf4 committed:
Incremental commit to aid in reviewing

bf4 commented:
rebased

bf4 commented:
@ethanresnick LGTM?

ethanresnick commented:
LGTM. Merged.

bf4 commented:
omg omg :fireworks: :rainbow: :exclamation: :boom: :dancers: :heart: :heart_eyes: :smiley: :100:

joaomdmoura commented:
:tada: :tada: :tada: :tada: :tada:

dgeb commented:
Really pleased to see this merged. We've needed some error examples for months.

@bf4 thanks for all your work here! :100: 

@ethanresnick thanks for finishing the review! :tada:
